tests/qnodes/test_qnode_passthru.py

```diff
@@ -0,0 +1,186 @@
+
+        node = PassthruQNode(circuit, tensornet_tf_device)
+        x = tf.Variable([1.1, 1.4], dtype=self.DTYPE)
+        res = node(x=x)
+        assert isinstance(res, tf.Tensor)
+        assert res.shape == (2,)
+        assert res.dtype == self.DTYPE
+
+    def test_tensor_operations(self, tensornet_tf_device, tol):
+        """Tests the evaluation of a PassthruQNode involving algebraic operations between tensor parameters,
+        and TF functions acting on them."""
+
+        def circuit(phi, theta):
+            x = phi * theta
+            qml.RX(x[0], wires=0)
+            qml.RY(tf.cos(x[1]), wires=1)
+            qml.CNOT(wires=[0, 1])
+            return qml.expval(qml.PauliZ(0)), qml.expval(qml.Hadamard(1))
+
+        node = PassthruQNode(circuit, tensornet_tf_device)
+
+        phi = tf.Variable(np.array([0.7, -1.2]), dtype=self.DTYPE)
+        theta = tf.Variable(1.7, dtype=self.DTYPE)
+        res = node(phi, theta)
+        assert isinstance(res, tf.Tensor)
+        assert res.shape == (2,)
+        assert res.dtype == self.DTYPE
+
+    def test_evaluate(self, tensornet_tf_device, tol):
+        """Tests the evaluation of a PassthruQNode."""
+
+        def circuit(x):
+            qml.RX(x[0], wires=0)
+            qml.RY(x[1], wires=1)
+            qml.CNOT(wires=[0, 1])
+            qml.RX(x[2], wires=1)
+            return qml.expval(qml.PauliZ(0)), qml.expval(qml.PauliZ(1))
+
+        x = np.array([-1.4, 0.2, 2.1])
+        x_tf = tf.Variable(x, dtype=self.DTYPE)
+
+        node = PassthruQNode(circuit, tensornet_tf_device)
+        res = node(x_tf)
+
+        # compare to a reference node
+        ref_device = qml.device('default.qubit', wires=2)
+        ref_node = BaseQNode(circuit, ref_device)
+        ref_res = ref_node(x)
+
+        assert isinstance(res, tf.Tensor)
+        assert res.shape == (2,)
+        assert res.dtype == self.DTYPE
+        assert res.numpy() == pytest.approx(ref_res, abs=tol)
+
+    @pytest.mark.parametrize("vectorize_jacobian", [True, False])
+    def test_jacobian(self, tensornet_tf_device, vectorize_jacobian, tol):
+        """Tests the computing of the Jacobian of a PassthruQNode using TensorFlow."""
+
+        def circuit(phi, theta):
+            qml.RX(phi[0], wires=0)
+            qml.RY(phi[1], wires=1)
+            qml.CNOT(wires=[0, 1])
+            qml.PhaseShift(theta, wires=0)
+            return qml.expval(qml.PauliZ(0)), qml.expval(qml.Hadamard(1))
+
+        ph = np.array([0.7, -1.2])
+        th = 1.7
+        phi = tf.Variable(ph, dtype=self.DTYPE)
+        theta = tf.Variable(th, dtype=self.DTYPE)
+
+        node = PassthruQNode(circuit, tensornet_tf_device)
+        # In TF 2, tf.GradientTape.jacobian comes with a vectorization option.
+        with tf.GradientTape(persistent=not vectorize_jacobian) as tape:
+            tape.watch([phi, theta])
+            res = node(phi, theta)
+        phi_grad, theta_grad = tape.jacobian(res, [phi, theta],
```

Comment:
There seems to be a few non-straightforward things going together here (e.g., `tf.UnconnectedGradients.ZERO`, `experimental_use_pfor`, etc.). Is there not a simpler "unit" test case we can do first?

Reply:
`tf.UnconnectedGradients.ZERO` simply asks for an output which has zeros if the result cannot depend on the parameter, instead of Nones.

The test is parametrized over the `vectorize_jacobian` option provided by `tape.jacobian`, but we could just test one of the two methods.

https://www.tensorflow.org/api_docs/python/tf/GradientTape#jacobian

Reply:
Got it, thanks